src/training/alpha_zero_trainer.py

```diff
@@ -0,0 +1,230 @@
+from multiprocessing import Pool
+from ..models import alpha_zero_net as Net
+from .trainer import Trainer
+from ..games import *
+
+import random
+import time
+import os
+import json
+
+class AlphaZeroTrainer(Trainer):
+    """
+    Trainer manager for Alpha Zero model
+    """
+    def __init__(
+        self,
+        net: Net,
+        batch_size: int,
+        handouts: int,
+        rollouts: int,
+        max_number: int,
+        pieces_per_player: int,
+        data_path: str = 'data'
+    ):
+        """
+        param net: nn.Module
+            Neural Network to train
+        param batch_size: int
+            Size of training data used per epoch
+        param handouts: int
+            Number of handouts per move search
+        param rollouts: int
+            Number of rollouts per handout
+        param max_number:
+            Max piece number
+        param pieces_per_player:
+            Number of pieces distributed per player
+        param data_path: string
+            Path to the folder where training data will be saved
+        """
+        self.net = net
+        self.batch_size = batch_size
+        self.handouts = handouts
+        self.rollouts = rollouts
+        self.max_number = max_number
+        self.pieces_per_player = pieces_per_player
+        self.data_path = data_path
+        self.error_log = []
+
+        if not os.path.exists(self.data_path):
+            os.makedirs(self.data_path)
+
+        self.net.eval()
+        
+    def self_play(
+        self,
+        handouts,
+        rollouts
+    ):
+        """
+        Simulate one game via self play and save moves played
+
+        param handouts: int
+            Number of handouts per move search
+        param rollouts: int
+            Number of rollouts per handout
+
+        return
+            Data of the game. List[(state, pi, result)]
+        """
+        data = []
+        game_over = False
+        root = True
+        players = [BasePlayer(i) for i in range(4)]
+        manager = DominoManager()
+        manager.init(players, hand_out, self.max_number, self.pieces_per_player)
+
+        while not game_over:
+            stats = {}
+            cur_player = players[manager.domino.current_player]
+            selector = selector_maker(stats, cur_player.valid_moves(), cur_player.pieces_per_player - len(cur_player.pieces), root, 6)
```

Comment:
also we should parameterize this number(6)

Reply:
Yes, that can be done